path/path.go

```diff
@@ -3,187 +3,240 @@ package path
+}
 
-	// if the path doesnt begin with a '/'
-	// we expect this to start with a hash, and be an 'ipfs' path
-	if parts[0] != "" {
-		if _, err := decodeCid(parts[0]); err != nil {
-			return "", &ErrInvalidPath{error: err, path: txt}
-		}
-		// The case when the path starts with hash without a protocol prefix
-		return Path("/ipfs/" + txt), nil
+func NewImmutablePath(p Path) (ImmutablePath, error) {
+	if p.Namespace().Mutable() {
+		return nil, ErrInvalidPath{err: ErrExpectedImmutable, path: p.String()}
 	}
 
-	if len(parts) < 3 {
-		return "", &ErrInvalidPath{error: fmt.Errorf("invalid ipfs path"), path: txt}
+	segments := p.Segments()
+	cid, err := cid.Decode(segments[1])
+	if err != nil {
+		return nil, &ErrInvalidPath{err: err, path: p.String()}
 	}
 
-	// TODO: make this smarter
-	switch parts[1] {
-	case "ipfs", "ipld":
-		if parts[2] == "" {
-			return "", &ErrInvalidPath{error: fmt.Errorf("not enough path components"), path: txt}
-		}
-		// Validate Cid.
-		_, err := decodeCid(parts[2])
-		if err != nil {
-			return "", &ErrInvalidPath{error: fmt.Errorf("invalid CID: %w", err), path: txt}
-		}
-	case "ipns":
-		if parts[2] == "" {
-			return "", &ErrInvalidPath{error: fmt.Errorf("not enough path components"), path: txt}
-		}
-	default:
-		return "", &ErrInvalidPath{error: fmt.Errorf("unknown namespace %q", parts[1]), path: txt}
-	}
+	return immutablePath{path: p, cid: cid}, nil
+}
 
-	return Path(txt), nil
+func (ip immutablePath) String() string {
+	return ip.path.String()
 }
 
-// ParseCidToPath takes a CID in string form and returns a valid ipfs Path.
-func ParseCidToPath(txt string) (Path, error) {
-	if txt == "" {
-		return "", &ErrInvalidPath{error: fmt.Errorf("empty"), path: txt}
-	}
+func (ip immutablePath) Namespace() Namespace {
+	return ip.path.Namespace()
+}
 
-	c, err := decodeCid(txt)
-	if err != nil {
-		return "", &ErrInvalidPath{error: err, path: txt}
-	}
+func (ip immutablePath) Segments() []string {
+	return ip.path.Segments()
+}
 
-	return FromCid(c), nil
+func (ip immutablePath) Cid() cid.Cid {
+	return ip.cid
 }
 
-// IsValid checks if a path is a valid ipfs Path.
-func (p *Path) IsValid() error {
-	_, err := ParsePath(p.String())
-	return err
+func (ip immutablePath) Remainder() string {
+	remainder := strings.Join(ip.Segments()[2:], "/")
+	if remainder != "" {
+		remainder = "/" + remainder
+	}
+	return remainder
 }
 
-// Join joins strings slices using /
-func Join(pths []string) string {
-	return strings.Join(pths, "/")
+// NewIPFSPath returns a new "/ipfs" path with the provided CID.
+func NewIPFSPath(cid cid.Cid) ImmutablePath {
+	return immutablePath{
+		path: path{
+			str:       fmt.Sprintf("/%s/%s", IPFSNamespace, cid.String()),
+			namespace: IPFSNamespace,
+		},
+		cid: cid,
+	}
 }
 
-// SplitList splits strings usings /
-func SplitList(pth string) []string {
-	return strings.Split(pth, "/")
+// NewIPLDPath returns a new "/ipld" path with the provided CID.
+func NewIPLDPath(cid cid.Cid) ImmutablePath {
```

Comment:
I'm not sure I understand the point of this function. This probably means we need more description around what the IPFS and IPLD namespaces are for.

As I understand it:
- `/ipfs` -> UnixFS pathing semantics with fallback to IPLD data model pathing semantics. Don't recall expectations around empty path segments or awkward characters (e.g. `NUL`).
- `/ipld` -> Just IPLD data model pathing. AFAIK this is form of path has been rarely used so far. Lack of definition for where ADLs, etc. fit in has been rough here. It's also not really defined anywhere so it's in a bit of limbo.

Are there supposed to be semantics for what happens at the end of the path with `/ipfs` and `/ipld` for this interface? Are we defining a resource or just how to get to the terminal element and then what happens with that resource is up to the consumer? From what I can tell it looks like the latter, although at some point we may also need one for the former (e.g. that could include things like byte ranges, IPLD ADLs, etc.).

I understand this mess is already here and we may not want to broach this particular mess right now, but even if not it's probably worth some documentation to help clarify the state of things and point to a GitHub issue for further discussion if people run into problems.

Reply:
> I'm not sure I understand the point of this function.

It just creates a new path under the `/ipld` namespace with that CID, so `/ipld/{cid}`. I just kept it because we're using it in other parts of the code.

> From what I can tell it looks like the latter

Yes.

> I understand this mess is already here and we may not want to broach this particular mess right now, but even if not it's probably worth some documentation to help clarify the state of things and point to a GitHub issue for further discussion if people run into problems.

Yeah, I'm not sure if we should tackle this now. I think it also relates to your other comment about [pathing through empty names](https://github.com/ipfs/boxo/pull/334#discussion_r1315988175) (I cannot reply to that comment, I don't know why).

> https://github.com/ipfs/boxo/pull/334#discussion_r1315988175

I think that is a valid concern. My problem is this is the following: before, we were not "cleaning" the path when creating a new path, or when validating it. However, when we split it into segments we were: https://github.com/ipfs/boxo/blob/574df963653c940ea19b1b8a0d37cd30360c22e9/path/path.go#L36 - which I would argue would give an inconsistent result. I don't think that it is very likely that we will want to path through empty namespaces, or namespaces that get collapsed by `gopath.Clean`, such as `.` and `..`. If we do, we will definitely need to update this and maybe have a different type for web paths (cc @lidel as there is some issue for specs for that), unixfs paths, and whatnot. I would argue that at the moment there's no real use case for the things that get collapsed/cleaned by `gopath.Clean`.

Reply:
Removed, and `FromCid` restored.